doc/neps/indexing.rst

```diff
@@ -0,0 +1,506 @@
+   be possible to be made more explicitly (e.g. ``arr.vindex[indices]``).
+
+3. A new ``arr.vindex[indices]`` method, would not be tied to the
+   confusing transpose rules of fancy indexing, which is for example
+   needed for the simple case of a single advanced index. Thus,
+   no transposing should be done. The axes created by the integer array
+   indices are always inserted at the front, even for a single index.
+
+4. Boolean indexing is conceptionally outer indexing. A broadcasting
+   together with other advanced indices in the manner of legacy
+   "fancy indexing" is generally not helpful or well defined.
+   A user who wishes the "``nonzero``" plus broadcast behaviour can thus
+   be expected to do this manually. Thus, ``vindex`` would still use
+   outer type indexing for boolean index arrays.
+   Note that using this rule a single boolean index can still index into
+   multiple dimensions at once.
+
+5. An ``arr.lindex`` or ``arr.findex`` should likely be implemented to allow
+   legacy fancy indexing indefinetly. This also gives a simple way to
+   update fancy indexing code making deprecations to plain indexing
+   easier.
+
+6. Plain indexing ``arr[...]`` should return an error for ambiguous cases.
+   For the beginning, this probably means cases where ``arr[ind]`` and
+   ``arr.oindex[ind]`` return different results give deprecation warnings.
+   Due to the transposing behaviour, this means that``arr[0, :, index_arr]``
+   will be deprecated, but ``arr[:, 0, index_arr]`` will not for the time
+   being.
+
+Unlike plain indexing, the new indexing attributes are explicitly aimed
+at higher dimensional indexing, two more changes should be implemented:
+
+* The indexing attributes will enforce exact dimension and indexing match.
+  This means that no implicit ellipsis (``...``) will be added. Unless
+  an ellipsis is present the indexing expression will thus only work for
+  an array with a specific number of dimensions.
+  This makes the expression more explicit and safeguards against wrong
+  dimensionality of arrays.
+
+* The current plain indexing allows for the use of non-tuples for
+  multi-dimensional indexing such as ``arr[[slice(None), 2]]``.
+  This creates some inconsistencies and thus the indexing attributes
+  should only allow plain python tuples for this purpose.
+  (Whether or not this should be the case for plain indexing is a
+  different issue.)
+
+* The new attributes should not do the getitem to implement setitem
+  branch, since it is a cludge and not useful for vectorized
+  indexing. (not implemented yet)
+
+
+Open Questions
+==============
+
+* The names ``oindex`` and ``vindex`` are just suggestions at the time of
+  writing this, another name NumPy has used for something like ``oindex``
+  is ``np.ix_``. See also below.
+
+* It would be possible to limit the use of boolean indices in ``vindex``,
+  assuming that they are rare and to some degree special. Alternatively,
```

Comment:
For the sake of everyone's sanity, I propose making the simpler choice of forbidding boolean indices in `vindex`.

Reply:
Fine with me :).